### PR TITLE
forward loop for same height sorting

### PR DESCRIPTION
**Pull request by rgknox**

### CODE REVIEW

File: biogeochem/EDCanopyStructureMod.F90 [hunk 1]
```diff
@@ -59,6 +59,9 @@ module EDCanopyStructureMod
   real(r8), parameter :: area_target_precision = 1.0E-11_r8  ! Area conservation must be within this tolerance
   real(r8), parameter :: area_check_precision  = 1.0E-9_r8  ! Area conservation checks must be within this tolerance
 
+  real(r8), parameter :: similar_height_tol = 1.0E-3_r8    ! I think trees that differ by 1mm
```
Comment (ckoven) [inline biogeochem/EDCanopyStructureMod.F90:62]: i worry that this is too large a difference threshold.  during the demotion process, a cohort might get split and then the demoted one will lag in size by only 1 day's worth of growth from the cohort that remained in the canopy.  presumably that will be less than 1 mm.  so if we let this be big, we could re-mix the just demoted cohort back into the canopy?  i'd prefer we drop this value by a couple more orders of magnitude so as only do this in the case where things are really very equal.

File: biogeochem/EDCanopyStructureMod.F90 [hunk 2]
```diff
@@ -367,117 +370,97 @@ subroutine DemoteFromLayer(currentSite,currentPatch,i_lyr)
...
-                  endif
+               else
 
-                  if ( tied_size_with_neighbor ) then
-                     if ( DEBUG ) then
-                        write(fates_log(),*) 'tied_size_with_neighbor eq true in demotion phase'
+                  ! -----------------------------------------------------------
+                  ! Rank ordered deterministic method
+                  ! -----------------------------------------------------------
+                  ! If there cohorts that have the exact same height (which is possible, really)
```
Comment (ckoven) [inline biogeochem/EDCanopyStructureMod.F90]:
```suggestion
                  ! If there are cohorts that have the exact same height (which is possible, really)
```

File: biogeochem/EDCanopyStructureMod.F90 [hunk 3]
```diff
@@ -830,113 +814,87 @@ subroutine PromoteIntoLayer(currentSite,currentPatch,i_lyr)
...
+
                   if (ED_val_comp_excln .ge. 0.0_r8 ) then
                      ! normal (stochastic) case, as above.
                      currentCohort%prom_weight = currentCohort%n*currentCohort%hite**ED_val_comp_excln
+                     sumweights = sumweights + currentCohort%prom_weight
                   else
+                     
+                     ! ------------------------------------------------------------------
                      ! Rank ordered deterministic method
+                     ! If there cohorts that have the exact same height (which is possible, really)
```
Comment (ckoven) [inline biogeochem/EDCanopyStructureMod.F90]:
```suggestion
                     ! If there are cohorts that have the exact same height (which is possible, really)
```

File: biogeochem/EDCanopyStructureMod.F90 (hunk 1 above)
Comment (ckoven) [reply biogeochem/EDCanopyStructureMod.F90:62]: oh, wait.  i just realized that the logic prevents that situation from happening.  so maybe it is ok to leave it at 1mm.

**Review by ckoven — APPROVED** [review]
this all looks good to me.  much simpler logic than what i had written.  thanks @rgknox.

**Comment by rgknox** [comment]
cool, testing it now, looks like things are working, will check back

**Comment by rgknox** [comment]
@ckoven , these changes passed the fates regression test.
I also ran a 60 year f45 test using the 2 tropical clone parameter set, using the ranked exclusion mode.  The two pfts are not b4b exactly the same,  but I'm still in the process of determining how much different or similar they are.  If you want, feel free to merge this, and we can go ahead and commit to master.

**Comment by rgknox** [comment]
Here is the comparison of the two tropical PFTS biomass after 60 years on the f45 grid:
https://drive.google.com/file/d/1RnhBItuqArGeoWEyUpH4jL-JOzZ4AwX1/view?usp=sharing
So this method does not do such a great job at preserving identicalness in the two...
I think I will give your original method a try too.   Sorry if this turned into more complication than its worth.

**Comment by ckoven** [comment]
i think its good at least that there isn't a consistent winner.  so it does seem to just be generating numerical noise.  i guess the important part is that it behaves better than what's on master -- could you run the exact same test with that?

**Comment by rgknox** [comment]
master is kinda weird, it is really consistent, but it has a bias:
https://drive.google.com/file/d/144PRogHePpi9aeqnUllXTCIPJgBnfn8p/view?usp=sharing

**Comment by rgknox** [comment]
actually, it was just the color axis made it look like a bias.   master, using non-ranked, gives almost exactly the same biomass from 1 pft to the other.  See the last figure here:
https://drive.google.com/open?id=144PRogHePpi9aeqnUllXTCIPJgBnfn8p

**Comment by ckoven** [comment]
that is really weird.  lets chat when I'm back in town so that I can try to understand how this is the case.  thanks for running the simulation.

**Comment by rgknox** [comment]
I updated the plots, included master, the base changes you proposed, and then my changes. Both of our versions aren't quite as good as the non-ranked master version at maintaining the parity of clones.

master (excln =3) : https://drive.google.com/file/d/1tPu1Z_ECrDLow6DhlDd5_c7UlykZexQW/view?usp=sharing

charlies PR version  (excln = -1) : https://drive.google.com/file/d/1PYEv98qUgva6b9Uqw8QTnrYMx5rmbrIO/view?usp=sharing

ryan's "updates" (excln = -1) : 
https://drive.google.com/file/d/1veQuwBTnii0f5d8oENADPLEdkd72f2Oc/view?usp=sharing

excln = 3, but with new height based sorting:
https://drive.google.com/file/d/1t_J6n5d-0N8wMCblkwCc06M8Y5n3ikXQ/view?usp=sharing

**Comment by rgknox** [comment]
I'm wondering if this has more to do with height sorting instead of dbh sorting... I'm now running a test to see if the probabalistic exclusion using height based sorting preserves clones as well as master was.

**Comment by ckoven** [comment]
Could you do a master (excln =-1) run?  That is the minimum bar we need to exceed here.

**Comment by ckoven** [comment]
ps your version actually look pretty good!

**Comment by rgknox** [comment]
I posted a new version, it is excln = 3, with height based sorting.  With height based sorting, there are now grid-cells that show divergence between the two clones that were not shown in the dbh sorting version.  However, the differences are still much smaller than what we see in the two ranked promotion/demotion tests.  

Although, I think it partially explains why we found that ranked prom/demo was not quite as good as the probabalistic prom/demo WRT preserving clones, what was really at play was that its harder to do when we are using heights instead of DBH.

master with excln = -1, I could run it through tests, but I think it just isn't coded to preserve clone similarity at all, there is no doubt there will be divergence. Although, I guess it would give us a sense of what the baseline is.

**Comment by ckoven** [comment]
ok, cool. no need to do the master/-1 case.  agreed that that was only meant to be a minimal point.  looks to me like your changes are a lot better than what I had, so I'm happy to pull this if you feel its ready.  then, separately, we should figure out why height sorting isn't giving identical cohorts, since in principle it still should.

**Comment by rgknox** [comment]
great. If you merge this in, I will begin integration with master.

**Pull request merged**